ray-operator/controllers/ray/common/route_test.go

```diff
@@ -0,0 +1,73 @@
+package common
+
+import (
+	rayv1alpha1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+
+	"strings"
+	"testing"
+)
+
+var instanceWithRouteEnabled = &rayv1alpha1.RayCluster{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "raycluster-sample",
+		Namespace: "default",
+		Annotations: map[string]string{
+			IngressClassAnnotationKey: "nginx",
+		},
+	},
+	Spec: rayv1alpha1.RayClusterSpec{
+		RayVersion: "1.0",
+		HeadGroupSpec: rayv1alpha1.HeadGroupSpec{
+			Replicas:      pointer.Int32Ptr(1),
+			EnableIngress: pointer.BoolPtr(true),
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{
+						{
+							Name:    "ray-head",
+							Image:   "rayproject/autoscaler",
+							Command: []string{"python"},
+							Args:    []string{"/opt/code.py"},
+						},
+					},
+				},
+			},
+		},
+	},
+}
+
+func TestBuildRouteForHeadService(t *testing.T) {
+	route, err := BuildRouteForHeadService(*instanceWithRouteEnabled)
```

Comment:
Although currently `BuildRouteForHeadService` will only read `instanceWithRouteEnabled`, it is better to perform a deep copy of `instanceWithRouteEnabled` to avoid side effects in the future. I expect that `instanceWithRouteEnabled` will be used in multiple tests. If the value will be different after each test, it will prevent the tests from covering the correct code paths.

Reply:
Huh? `BuildRouteForHeadService` should never modify source

Reply:
In general, I would agree that shallow copy is bad. In this particular case `BuildRouteForHeadService` is part of CR processing. If it modifies CR we have a much bigger problem. Also compare it with other tests, for example ingress test

Reply:
@blublinsky, perhaps inlining the variable in the call to `BuildRouteForHead` might be a better solution. I think it should assuade @kevin85421 concerns and also slighly simplify the reading of the code (a subjective opinion)